common/blockchain/src/manager.rs

```diff
@@ -199,26 +199,27 @@ impl<PK: crate::PubKey> ChainManager<PK> {
         &mut self,
         pubkey: PK,
         signature: &PK::Signature,
+        verifier: &impl crate::Verifier<PK>,
     ) -> Result<bool, AddSignatureError> {
         let pending = self
             .pending_block
             .as_mut()
             .ok_or(AddSignatureError::NoPendingBlock)?;
-        if pending.signers.contains(&pubkey) {
-            return Ok(false);
-        }
-        if !pubkey.verify(pending.hash.as_slice(), signature) {
-            return Err(AddSignatureError::BadSignature);
-        }
-
-        pending.signing_stake += self
+        let validator_stake = self
             .next_epoch
             .validator(&pubkey)
             .ok_or(AddSignatureError::BadValidator)?
             .stake()
             .get();
-        assert!(pending.signers.insert(pubkey));
+        if !verifier.verify(pending.hash.as_slice(), &pubkey, signature) {
+            return Err(AddSignatureError::BadSignature);
+        }
+
+        if !pending.signers.insert(pubkey) {
+            return Ok(false);
```

Comment:
what does `Ok(false)` signals?

Reply:
/// Returns `true` if quorum has been reached and the pending block has
    /// graduated to the current block.

`Ok(false)` means that the signature was correct but the block is
still pending because quorum hasn’t been reached.